src/optimagic/optimizers/pygad_optimizer.py

```diff
@@ -0,0 +1,292 @@
+import warnings
+from dataclasses import dataclass
+from typing import Any, Literal
+
+import numpy as np
+from numpy.typing import NDArray
+
+from optimagic import mark
+from optimagic.config import IS_PYGAD_INSTALLED
+from optimagic.exceptions import NotInstalledError
+from optimagic.optimization.algo_options import get_population_size
+from optimagic.optimization.algorithm import Algorithm, InternalOptimizeResult
+from optimagic.optimization.internal_optimization_problem import (
+    InternalOptimizationProblem,
+)
+from optimagic.typing import (
+    AggregationLevel,
+    CrossoverFunction,
+    GeneConstraintFunction,
+    MutationFunction,
+    NonNegativeFloat,
+    ParentSelectionFunction,
+    PositiveFloat,
+    PositiveInt,
+)
+
+if IS_PYGAD_INSTALLED:
+    import pygad
+
+
+@mark.minimizer(
+    name="pygad",
+    solver_type=AggregationLevel.SCALAR,
+    is_available=IS_PYGAD_INSTALLED,
+    is_global=True,
+    needs_jac=False,
+    needs_hess=False,
+    needs_bounds=True,
+    supports_parallelism=True,
+    supports_bounds=True,
+    supports_infinite_bounds=False,
+    supports_linear_constraints=False,
+    supports_nonlinear_constraints=False,
+    disable_history=False,
+)
+@dataclass(frozen=True)
+class Pygad(Algorithm):
+    population_size: PositiveInt | None = None
+    num_parents_mating: PositiveInt | None = 10
+    num_generations: PositiveInt | None = 50
+
+    initial_population: NDArray[np.float64] | list[list[float]] | None = None
+
+    parent_selection_type: (
+        Literal["sss", "rws", "sus", "rank", "random", "tournament"]
+        | ParentSelectionFunction
+    ) = "sss"
+    keep_parents: int = -1
+    keep_elitism: PositiveInt = 1
+    K_tournament: PositiveInt = 3
+
+    crossover_type: (
+        Literal["single_point", "two_points", "uniform", "scattered"]
+        | CrossoverFunction
+        | None
+    ) = "single_point"
+    crossover_probability: NonNegativeFloat | None = None
+
+    mutation_type: (
+        Literal["random", "swap", "inversion", "scramble", "adaptive"]
+        | MutationFunction
+        | None
+    ) = "random"
+    mutation_probability: (
+        NonNegativeFloat
+        | list[NonNegativeFloat]
+        | tuple[NonNegativeFloat, NonNegativeFloat]
+        | NDArray[np.float64]
+        | None
+    ) = None
+    mutation_percent_genes: (
+        PositiveFloat
+        | str
+        | list[PositiveFloat]
+        | tuple[PositiveFloat, PositiveFloat]
+        | NDArray[np.float64]
+    ) = "default"
+    mutation_num_genes: (
+        PositiveInt
+        | list[PositiveInt]
+        | tuple[PositiveInt, PositiveInt]
+        | NDArray[np.int_]
+        | None
+    ) = None
+    mutation_by_replacement: bool = False
+    random_mutation_min_val: float | list[float] | NDArray[np.float64] = -1.0
+    random_mutation_max_val: float | list[float] | NDArray[np.float64] = 1.0
```

Comment:
In the array case: Does the array have one entry per parameter? If so, we would need o more general PyTree handling with the converter here.